### PR TITLE
Add restoring event listener

### PR DESCRIPTION
**Pull request by heemin32**
### Description
In the listener, we trigger a geoip data update
 
### Issues Resolved
N/A
 
### Check List
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/Datasource.java [hunk 1]
```diff
@@ -537,7 +553,7 @@ public Database(final StreamInput in) throws IOException {
         public void writeTo(final StreamOutput out) throws IOException {
             out.writeOptionalString(provider);
             out.writeOptionalString(sha256Hash);
-            out.writeOptionalVLong(updatedAt.toEpochMilli());
+            out.writeOptionalVLong(updatedAt == null ? null : updatedAt.toEpochMilli());
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/Datasource.java:556]: isn't the check for null already encapsulated into method if it does "optional"?

File: src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/Datasource.java [hunk 2]
```diff
@@ -640,10 +656,10 @@ public UpdateStats(final StreamInput in) throws IOException {
 
         @Override
         public void writeTo(final StreamOutput out) throws IOException {
-            out.writeOptionalVLong(lastSucceededAt.toEpochMilli());
+            out.writeOptionalVLong(lastSucceededAt == null ? null : lastSucceededAt.toEpochMilli());
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/Datasource.java:659]: same as for previous call to write optional, I think null value should be already handled

File: src/main/java/org/opensearch/geospatial/ip2geo/listener/Ip2GeoListener.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import org.opensearch.jobscheduler.spi.schedule.IntervalSchedule;
+import org.opensearch.threadpool.ThreadPool;
+
+@Log4j2
+public class Ip2GeoListener extends AbstractLifecycleComponent implements ClusterStateListener {
+    private static final int SCHEDULE_IN_MIN = 15;
+    private static final int DELAY_IN_MILLIS = 10000;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/geospatial/ip2geo/listener/Ip2GeoListener.java:37]: nit: extra empty line

File: src/main/java/org/opensearch/geospatial/ip2geo/listener/Ip2GeoListener.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    private final DatasourceFacade datasourceFacade;
+
+    @Inject
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/geospatial/ip2geo/listener/Ip2GeoListener.java:40]: curious, is this different/can be replaced by lombok's AllArgsConstructor?

File: src/main/java/org/opensearch/geospatial/ip2geo/listener/Ip2GeoListener.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    @Override
+    protected void doStop() {
+        clusterService.removeListener(this);
+    }
+
+    @Override
+    protected void doClose() throws IOException {
+
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/geospatial/ip2geo/listener/Ip2GeoListener.java:116]: should we do same as on stop here?

File: src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/Datasource.java (hunk 1 above)
Comment (heemin32) [reply src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/Datasource.java:556]: Here I call `.toEpochMillis()` method and will will throw NPE if I don't do null check.

File: src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/Datasource.java (hunk 2 above)
Comment (heemin32) [reply src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/Datasource.java:659]: Same for previous one. I call `.toEpochMillis()` method.

File: src/main/java/org/opensearch/geospatial/ip2geo/listener/Ip2GeoListener.java (hunk 5 above)
Comment (heemin32) [reply src/main/java/org/opensearch/geospatial/ip2geo/listener/Ip2GeoListener.java:116]: I think `doStop` is called always before `doClose()`. Removing listener in `doClose()` will have no effect.
Saw the same pattern in OpenSearch core.

**Review by martin-gaievski — APPROVED** [review]

**Review by junqiu-lei — APPROVED** [review]

**Pull request merged**